### ISSUE TITLE
Numerical overflow when using contextual metrics 

### ISSUE THREAD

**Issue by kronerte** [signals-dev/Orion#212]
* Orion version: 0.1.5
* Python version: 3.7.10
* Operating System: windows 10

### Description

I try to reproduce the results from the notebook: Pipeline Scoring Example.ipynb from the github repository.
This notebook evaluate the f1 score of the pipeline lstm_dynamic_threshold on the S-1 timeserie.

When I run the notebook, whithout changing anything, it works whithout raising any exception.

But, he resulting F-1 score is very different from what is expected from the github result.

It seams that I have such an instability issue.

### What I Did

When I run the notebook, the final F-1 score I get is:
![image](https://user-images.githubusercontent.com/30753759/111225764-79595c00-85e0-11eb-8d03-e63fe52aa21c.png)

But according to github, the results should be:
![image](https://user-images.githubusercontent.com/30753759/111226073-e2d96a80-85e0-11eb-9ab4-b0f412cb03c0.png)

So there is a huge difference, I don't know from where it could come.
When I rerun the notebook, I get differents results: the f1 score takes values from: 0.08 to 0.3.

To give a little more details:

According to the logs of the training, it seems not to have huge difference with the github source:
own results:
![image](https://user-images.githubusercontent.com/30753759/111226670-c2f67680-85e1-11eb-8123-a75dfe4d563e.png)
result from orion's gihub:
![image](https://user-images.githubusercontent.com/30753759/111226776-e15c7200-85e1-11eb-9bce-4f1ee647bad8.png)

But the anomalies I detect is a longer sequence, so that explain the precision and the recall I get is worse:
own result:
![image](https://user-images.githubusercontent.com/30753759/111227048-47e19000-85e2-11eb-89e4-a736de6bb751.png)
result from orion's gihub:
![image](https://user-images.githubusercontent.com/30753759/111227099-5af46000-85e2-11eb-8a7d-f541e396b8e9.png)

**Comment by sarahmish** [comment]
Hi @kronerte 

Thanks for posting this issue! It's typical for deep learning methods to produce different results in each run since the model parameters are randomly initialized. Therefore, the detected anomalous region might deviate from one experiment to another and that's the case you see here. Nevertheless, in both instances shown the same region was detected but with different intervals.

**How does this affect f1 score?**
To answer this question, we need to look at how we compute the f1 score. A step by step process is explained [here](https://github.com/signals-dev/Orion/tree/master/orion/evaluation). Basically, let's view the ground truth anomaly `known_anomalies`:
```
             start             end
0	1398168000	1407823200
```
in the first run we detected the region `anomalies_orion`:
```
             start             end         score
0	1398600000	1406030400	0.434961
```
and in the second run we detected the region `anomalies_yours`:
```
             start             end         score
0	1398556800	1406332800	0.488983
```
I notice that `anomalies_yours` covers a larger portion of the `known_anomalies`, therefore it should have a higher score.. If I use these definitions, this is the result I got

```python3
In   [1]: contextual_f1_score(known_anomalies, anomalies_orion, test)
out  [1]: 0.8697

In   [2]: contextual_f1_score(known_anomalies, anomalies_yours, test)
out  [2]: 0.8921
```
I'm thinking the `anomalies_yours` used to obtained an f1 score of `0.0814` shown previously is from another run. 

In short, yes results will differ from one to another, and that will be reflected in the evaluation score. I hope this answers your question!

**Comment by kronerte** [comment]
Hi @sarahmish

Thank you for your response.
That help me to specify the root of the problem and find a solution to it.
There is some variability on deep learning algorithm but it should not explain the difference between my score of `0.089` with the expected `0.89`.  

The issue come from a little incompatibility, I found, of the orion library when running on Windows.

## Weird behavior 

When I try to compute the f1 score for `anomalies_yours` on my local computer (Windows) I get:
```
[In]: contextual_f1_score([(1398168000, 1407823200)], [(1398600000, 1406030400)], test)
[out]: 0.0896533865
```

But when I run the same algorithm but with specifying `start` and `end` instead of `test`,
I get:
```
[In]: contextual_f1_score([(1398168000, 1407823200)], [(1398600000, 1406030400)], start=283688000, end=1442016000))
[out]: 0.8921
```
I also get the result `0.8921` when running the first command on `binder` server (Linux) instead of my local machine.

(The version of orion and Python are the same)

So I investigate on these issue, and I find the root of the problem.

## Root of the problem
### int32 vs int 64
When using `load_signal` the timestamp are not loaded on the same type in `Windows` and in `Linux`.

Indeed on the body of the  `load_signal` function we have:
```
data['timestamp'] = data['timestamp'].astype(int)
```
Doing so the type of the timestamp column is cast to the default integer of Numpy (or Pandas).

The default Integer in Numpy Correspond to the `C long` type.
https://www.tutorialspoint.com/numpy/numpy_data_types.htm

But `C long`  do not have the same size in Windows and in Linux.
In Windows  `C long`  have size 32bits (LLP64) and in Linux 64 bits (LP64).
https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models

The consequence is that in Linux the `timestamp` has a type `np.int64` and in Windows `np.int32`

### Consequences on the computation of f1 score
An integer overflow arrive on the function `_overlap`:
```
def _overlap(expected, observed):
    first = expected[0] - observed[1]
    second = expected[1] - observed[0]
    return first * second < 0
```
The variables   `first` and  `second` are "big" integer. If they are encoded on a  `np.int32` variables the product  ` first * second` can raise an integer overflow. Then the result of the product can have a wrong sing.

The error in overlap test computation propagate to the computation of the f1 score.

### With start and end

When we specify the `start` and `end` variables in `contextual_f1_score` function, the algorithm works with `int64`.
It seams that the default integer type for Python is  `int64`. (I am not sure)

## How I solved my issue
I solved my issue by casting the time of `test['timestamp']` to 'int64'
 ```
test['timestamp'] = test['timestamp'].astype('int64')
```

Then I get:
```
[In]: contextual_f1_score([(1398168000, 1407823200)], [(1398600000, 1406030400)], start=283688000, end=1442016000))
[out]: 0.8921
```
## Suggestion
In order to improve the compatibility of the library with Windows computer, I would suggest a modification in the code of the   `load_signal` function.

By specifying explicitly that the `timestamp` is on `int64`

So replacing:
 ```
data['timestamp'] = data['timestamp'].astype(int)
```
By:
```
data['timestamp'] = data['timestamp'].astype('int64')
```

**Comment by sarahmish** [comment]
Awesome, thank you @kronerte for the detailed report! 

I would invite you to make a pull request with your suggestion to give credit where credit is due. If you prefer not to, no problem, I can also make this modification :)

**Comment by kronerte** [comment]
Ok, I made the pull request.

So, I close this issue.
